File: Systems-for-Design-and-Implementation/Labs/Lab9/ProiectJava/TransportNetworking/src/main/java/transport/network/jsonprotocol/TransportClientJsonWorker.java
package transport.network.jsonprotocol;

import com.google.gson.*;
import transport.model.Reservation;
import transport.model.ReservationManager;
import transport.model.Seat;
import transport.model.Trip;
import transport.services.ITransportObserver;
import transport.services.ITransportServices;
import transport.services.TransportException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

public class TransportClientJsonWorker implements Runnable, ITransportObserver {
    private ITransportServices server;
    private Socket connection;

    private BufferedReader input;
    private PrintWriter output;
    private Gson gsonFormatter;
    private volatile boolean connected;

    public TransportClientJsonWorker(ITransportServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        gsonFormatter = new GsonBuilder()
                .registerTypeAdapter(LocalDateTime.class, (JsonSerializer<LocalDateTime>) (src, typeOfSrc, context) -> new JsonPrimitive(src.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)))
                .registerTypeAdapter(LocalDateTime.class, (JsonDeserializer<LocalDateTime>) (json, typeOfT, context) -> LocalDateTime.parse(json.getAsString(), DateTimeFormatter.ISO_LOCAL_DATE_TIME))
                .create();
        try {
            output = new PrintWriter(connection.getOutputStream());
            input = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void run() {
        while (connected) {
            try {
                String requestLine = input.readLine();
                Request request = gsonFormatter.fromJson(requestLine, Request.class);
                Response response = handleRequest(request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void reservationAdded(Reservation reservation) throws TransportException {
        Response response = JsonProtocolUtils.createNewReservationResponse(reservation);
        System.out.println("Reservation added " + reservation);
        try {
            sendResponse(response);
        } catch (IOException e) {
            throw new TransportException("Sending error: " + e);
        }
    }

    private static Response okResponse = JsonProtocolUtils.createOkResponse();

    private Response handleRequest(Request request) {
        //request.setType(RequestType.values()[request.getOrdinalType()]);
        Response response = null;
        if (request.getType() == RequestType.LOGIN) {
            System.out.println("Login request..." + request.getType());
            ReservationManager reservationManager = request.getReservationManager();
            try {
                server.login(reservationManager, this);
                return okResponse;
            } catch (TransportException e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.LOGOUT) {
            System.out.println("Logout request..." + request.getType());
            ReservationManager reservationManager = request.getReservationManager();
            try {
                server.logout(reservationManager, this);
                connected = false;
                return okResponse;
            } catch (TransportException e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.GET_ALL_TRIPS) {
            System.out.println("Get all trips request..." + request.getType());
            try {
                Trip[] trips = server.getAllTrips();
                System.out.println("Got all trips from Worker HandleRequest: ");
                for (Trip trip: trips)
                    System.out.println(trip);
                return JsonProtocolUtils.createGetAllTripsResponse(trips);
            } catch (TransportException e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.GET_RESERVATIONS_BY_TRIP) {
            System.out.println("Get reservations by trip request..." + request.getType());
            if (request.getTrip() == null)
                return JsonProtocolUtils.createErrorResponse("Trip in request is null");
            try {
                List<Reservation> reservations = server.getReservationsByTrip(request.getTrip());
                for (Reservation reservation: reservations)
                    System.out.println(reservation);
                return JsonProtocolUtils.createGetReservationsByTripResponse(reservations);
            } catch (TransportException e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.GET_SEATS_BY_TRIP) {
            System.out.println("Get seats by trip request..." + request.getType());
            if (request.getTrip() == null)
                return JsonProtocolUtils.createErrorResponse("Trip in request is null");
            try {
                List<Seat> seats = server.getSeatsByTrip(request.getTrip());
                for (Seat seat: seats)
                    System.out.println(seat);
                return JsonProtocolUtils.createGetSeatsByTripResponse(seats);
            } catch (TransportException e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.SEARCH_TRIP_BY_ID) {
            System.out.println("Search trip by id request..." + request.getType());
            if (request.getLong() == null)
                return JsonProtocolUtils.createErrorResponse("Trip id in request is null");
            try {
                Trip trip = server.searchTripById(request.getLong());
                return JsonProtocolUtils.createSearchTripByIdResponse(trip);
            } catch (TransportException e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.NEW_RESERVATION) {
            System.out.println("Add reservation request..." + request.getType());
            Reservation reservation = (Reservation) request.getReservation();
            try {
                server.addReservation(reservation);
                return okResponse;
            } catch (TransportException e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        //response.setOrdinalType(response.getType().ordinal());
        return response;
    }

    private void sendResponse(Response response) throws IOException {
        String responseLine = gsonFormatter.toJson(response);
        System.out.println("Sending response " + responseLine);
        synchronized (output) {
            output.println(responseLine);
            output.flush();
        }
    }
}

File: S4-Systems-for-Design-and-Impl/Labs/Lab8/ProiectCSharp/TransportNetworking/transport.network/jsonprotocol/TransportClientJsonWorker.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using Transport.Model.transport.model;
using TransportModel.transport.model;
using TransportServices.transport.services;

namespace TransportNetworking.transport.network.jsonprotocol
{
    public class TransportClientJsonWorker : ITransportObserver
    {
        private ITransportServices server;
        private Socket connection;

        private StreamReader input;
        private StreamWriter output;
        private JsonSerializerOptions options;
        private volatile bool connected;

        public TransportClientJsonWorker(ITransportServices server, Socket connection)
        {
            this.server = server;
            this.connection = connection;
            options = new JsonSerializerOptions
            {
                Converters = { new JsonStringEnumConverter() }
            };
            try
            {
                NetworkStream networkStream = new NetworkStream(this.connection);
                output = new StreamWriter(networkStream);
                input = new StreamReader(networkStream);
                connected = true;
            }
            catch (IOException e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public void Run()
        {
            while (connected)
            {
                try
                {
                    string requestLine = input.ReadLine();
                    Request request = JsonSerializer.Deserialize<Request>(requestLine, options);
                    Response response = HandleRequest(request);
                    if (response != null)
                    {
                        SendResponse(response);
                    }
                }
                catch (IOException e)
                {
                    Console.WriteLine(e.StackTrace);
                }
                Thread.Sleep(1000);
            }
            try
            {
                input.Close();
                output.Close();
                connection.Close();
            }
            catch (IOException e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public void ReservationAdded(Reservation reservation)
        {
            Response response = JsonProtocolUtils.CreateNewReservationResponse(reservation);
            Console.WriteLine("Reservation added " + reservation);
            try
            {
                SendResponse(response);
            }
            catch (IOException e)
            {
                throw new TransportException("Sending error: " + e);
            }
        }

        private static Response okResponse = JsonProtocolUtils.CreateOkResponse();

        private Response HandleRequest(Request request)
        {
            Response response = null;
            if (request.Type == RequestType.LOGIN)
            {
                Console.WriteLine("Login request..." + request.Type);
                ReservationManager reservationManager = request.ReservationManager;
                try
                {
                    server.Login(reservationManager, this);
                    return okResponse;
                }
                catch (TransportException e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }
            }
            if (request.Type == RequestType.LOGOUT)
            {
                Console.WriteLine("Logout request..." + request.Type);
                ReservationManager reservationManager = request.ReservationManager;
                try
                {
                    server.Logout(reservationManager, this);
                    connected = false;
                    return okResponse;
                }
                catch (TransportException e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }
            }
            if (request.Type == RequestType.GET_ALL_TRIPS)
            {
                Console.WriteLine("Get all trips request..." + request.Type);
                try
                {
                    Trip[] trips = server.GetAllTrips();
                    Console.WriteLine("Got all trips from Worker HandleRequest: ");
                    foreach (Trip trip in trips)
                        Console.WriteLine(trip);
                    return JsonProtocolUtils.CreateGetAllTripsResponse(trips);
                }
                catch (TransportException e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }
            }
            if (request.Type == RequestType.GET_RESERVATIONS_BY_TRIP)
            {
                Console.WriteLine("Get reservations by trip request..." + request.Type);
                if (request.Trip == null)
                    return JsonProtocolUtils.CreateErrorResponse("Trip in request is null");
                try
                {
                    List<Reservation> reservations = server.GetReservationsByTrip(request.Trip);
                    return JsonProtocolUtils.CreateGetReservationsByTripResponse(reservations);
                }
                catch (TransportException e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }
            }
            if (request.Type == RequestType.GET_SEATS_BY_TRIP)
            {
                Console.WriteLine("Get seats by trip request..." + request.Type);
                if (request.Trip == null)
                    return JsonProtocolUtils.CreateErrorResponse("Trip in request is null");
                try
                {
                    List<Seat> seats = server.GetSeatsByTrip(request.Trip);
                    foreach (Seat seat in seats)
                        Console.WriteLine(seat);
                    return JsonProtocolUtils.CreateGetSeatsByTripResponse(seats);
                }
                catch (TransportException e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }
            }
            if (request.Type == RequestType.SEARCH_TRIP_BY_ID)
            {
                Console.WriteLine("Search trip by id request..." + request.Type);
                if (request.Long == null)
                    return JsonProtocolUtils.CreateErrorResponse("Trip id in request is null");
                try
                {
                    Trip trip = server.SearchTripById(request.Long.Value);
                    return JsonProtocolUtils.CreateSearchTripByIdResponse(trip);
                }
                catch (TransportException e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }
            }
            if (request.Type == RequestType.NEW_RESERVATION)
            {
                Console.WriteLine("Add reservation request..." + request.Type);
                Reservation reservation = request.Reservation;
                try
                {
                    Reservation reservationAdded = server.AddReservation(reservation);
                    return okResponse;
                }
                catch (TransportException e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }
            }
            return response;
        }

        private void SendResponse(Response response)
        {
            string responseLine = JsonSerializer.Serialize(response, options);
            Console.WriteLine("Sending response " + responseLine);
            lock (output)
            {
                output.WriteLine(responseLine);
                output.Flush();
            }
        }
    }
}